File: iphone/ButtonLogTests/PushNotificationTests.swift
import XCTest
@testable import ButtonLog

/// Tests for push notification handling.
///
/// These tests verify:
/// - Notification payload parsing
/// - Notification handling logic
/// - Deep link extraction
/// - Badge count management
final class PushNotificationTests: XCTestCase {

    // MARK: - Notification Payload Parsing

    func testNotificationPayload_buttonClick_parsesCorrectly() {
        let payload: [String: Any] = [
            "aps": [
                "alert": [
                    "title": "Button Clicked",
                    "body": "John clicked 'Morning Routine'"
                ],
                "badge": 1,
                "sound": "default"
            ],
            "type": "button_click",
            "data": [
                "button_id": "btn-123",
                "button_name": "Morning Routine",
                "user_id": "user-456",
                "user_name": "John"
            ]
        ]

        let notification = parseNotificationPayload(payload)

        XCTAssertEqual(notification?.type, "button_click")
        XCTAssertEqual(notification?.title, "Button Clicked")
        XCTAssertEqual(notification?.body, "John clicked 'Morning Routine'")
        XCTAssertEqual(notification?.data?["button_id"] as? String, "btn-123")
    }

    func testNotificationPayload_friendRequest_parsesCorrectly() {
        let payload: [String: Any] = [
            "aps": [
                "alert": [
                    "title": "Friend Request",
                    "body": "Jane sent you a friend request"
                ],
                "badge": 2,
                "sound": "default"
            ],
            "type": "friend_request",
            "data": [
                "friend_id": "friend-789",
                "user_id": "user-456",
                "user_name": "Jane"
            ]
        ]

        let notification = parseNotificationPayload(payload)

        XCTAssertEqual(notification?.type, "friend_request")
        XCTAssertEqual(notification?.title, "Friend Request")
        XCTAssertEqual(notification?.data?["friend_id"] as? String, "friend-789")
    }

    func testNotificationPayload_systemMessage_parsesCorrectly() {
        let payload: [String: Any] = [
            "aps": [
                "alert": [
                    "title": "ButtonLog",
                    "body": "Your subscription has been renewed"
                ],
                "badge": 0,
                "sound": "default"
            ],
            "type": "system",
            "data": [
                "message_type": "subscription_renewed"
            ]
        ]

        let notification = parseNotificationPayload(payload)

        XCTAssertEqual(notification?.type, "system")
        XCTAssertEqual(notification?.data?["message_type"] as? String, "subscription_renewed")
    }

    // MARK: - Deep Link Extraction

    func testDeepLink_buttonClick_extractsCorrectly() {
        let notificationData: [String: Any] = [
            "button_id": "btn-123"
        ]

        let deepLink = extractDeepLink(type: "button_click", data: notificationData)

        XCTAssertEqual(deepLink, "buttonlog://button/btn-123")
    }

    func testDeepLink_friendRequest_extractsCorrectly() {
        let notificationData: [String: Any] = [
            "friend_id": "friend-456"
        ]

        let deepLink = extractDeepLink(type: "friend_request", data: notificationData)

        XCTAssertEqual(deepLink, "buttonlog://friends/requests")
    }

    func testDeepLink_unknownType_returnsHome() {
        let deepLink = extractDeepLink(type: "unknown", data: [:])

        XCTAssertEqual(deepLink, "buttonlog://home")
    }

    // MARK: - Badge Count Management

    func testBadgeCount_incrementsCorrectly() {
        var badgeCount = 0

        badgeCount = incrementBadge(currentCount: badgeCount)
        XCTAssertEqual(badgeCount, 1)

        badgeCount = incrementBadge(currentCount: badgeCount)
        XCTAssertEqual(badgeCount, 2)
    }

    func testBadgeCount_resetsToZero() {
        var badgeCount = 5

        badgeCount = resetBadge()
        XCTAssertEqual(badgeCount, 0)
    }

    func testBadgeCount_capsAtMaximum() {
        let maxBadge = 99
        var badgeCount = 98

        badgeCount = incrementBadge(currentCount: badgeCount, max: maxBadge)
        XCTAssertEqual(badgeCount, 99)

        badgeCount = incrementBadge(currentCount: badgeCount, max: maxBadge)
        XCTAssertEqual(badgeCount, 99) // Should not exceed max
    }

    // MARK: - Notification Grouping

    func testNotificationGrouping_buttonClicks() {
        let notifications: [ParsedNotification] = [
            ParsedNotification(type: "button_click", title: "Click 1", body: "", data: ["button_id": "btn-1"]),
            ParsedNotification(type: "button_click", title: "Click 2", body: "", data: ["button_id": "btn-1"]),
            ParsedNotification(type: "button_click", title: "Click 3", body: "", data: ["button_id": "btn-2"])
        ]

        let grouped = groupNotifications(notifications)

        // All three notifications have type "button_click", so they're grouped together
        XCTAssertEqual(grouped.count, 1)
        XCTAssertEqual(grouped["button_click"]?.count, 3)
    }

    func testNotificationGrouping_mixedTypes() {
        let notifications: [ParsedNotification] = [
            ParsedNotification(type: "button_click", title: "Click", body: "", data: [:]),
            ParsedNotification(type: "friend_request", title: "Request", body: "", data: [:]),
            ParsedNotification(type: "button_click", title: "Click 2", body: "", data: [:])
        ]

        let grouped = groupNotifications(notifications)

        XCTAssertEqual(grouped["button_click"]?.count, 2)
        XCTAssertEqual(grouped["friend_request"]?.count, 1)
    }

    // MARK: - Notification Permission State

    func testNotificationPermission_authorized_allowsNotifications() {
        let state = NotificationPermissionState.authorized

        XCTAssertTrue(canShowNotifications(state: state))
    }

    func testNotificationPermission_denied_blocksNotifications() {
        let state = NotificationPermissionState.denied

        XCTAssertFalse(canShowNotifications(state: state))
    }

    func testNotificationPermission_notDetermined_blocksNotifications() {
        let state = NotificationPermissionState.notDetermined

        XCTAssertFalse(canShowNotifications(state: state))
    }

    // MARK: - Device Token Handling

    func testDeviceToken_convertsToString() {
        // Simulated device token data (32 bytes)
        let tokenBytes: [UInt8] = [
            0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
            0x09, 0x0A, 0x0B, 0x0C, 0x0D, 0x0E, 0x0F, 0x10,
            0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 0x18,
            0x19, 0x1A, 0x1B, 0x1C, 0x1D, 0x1E, 0x1F, 0x20
        ]
        let tokenData = Data(tokenBytes)

        let tokenString = deviceTokenToString(tokenData)

        XCTAssertEqual(tokenString.count, 64) // 32 bytes = 64 hex chars
        XCTAssertTrue(tokenString.hasPrefix("01020304"))
    }

    // MARK: - Helper Methods

    private func parseNotificationPayload(_ payload: [String: Any]) -> ParsedNotification? {
        guard let aps = payload["aps"] as? [String: Any],
              let alert = aps["alert"] as? [String: Any],
              let title = alert["title"] as? String,
              let body = alert["body"] as? String else {
            return nil
        }

        let type = payload["type"] as? String ?? "unknown"
        let data = payload["data"] as? [String: Any]

        return ParsedNotification(type: type, title: title, body: body, data: data)
    }

    private func extractDeepLink(type: String, data: [String: Any]) -> String {
        switch type {
        case "button_click":
            if let buttonId = data["button_id"] as? String {
                return "buttonlog://button/\(buttonId)"
            }
        case "friend_request":
            return "buttonlog://friends/requests"
        default:
            break
        }
        return "buttonlog://home"
    }

    private func incrementBadge(currentCount: Int, max: Int = 99) -> Int {
        return min(currentCount + 1, max)
    }

    private func resetBadge() -> Int {
        return 0
    }

    private func groupNotifications(_ notifications: [ParsedNotification]) -> [String: [ParsedNotification]] {
        return Dictionary(grouping: notifications, by: { $0.type })
    }

    private func canShowNotifications(state: NotificationPermissionState) -> Bool {
        return state == .authorized
    }

    private func deviceTokenToString(_ data: Data) -> String {
        return data.map { String(format: "%02x", $0) }.joined()
    }
}

// MARK: - Data Types

struct ParsedNotification {
    let type: String
    let title: String
    let body: String
    let data: [String: Any]?
}

enum NotificationPermissionState {
    case notDetermined
    case denied
    case authorized
    case provisional
}

File: android/app/src/test/java/com/buttonlog/app/notifications/PushNotificationTests.kt
package com.buttonlog.app.notifications

import com.google.common.truth.Truth.assertThat
import com.google.gson.Gson
import org.junit.Test

/**
 * Tests for push notification handling.
 *
 * These tests verify:
 * - Notification payload parsing
 * - Notification handling logic
 * - Deep link extraction
 * - Badge count management
 */
class PushNotificationTests {

    private val gson = Gson()

    // MARK: - Notification Payload Parsing

    @Test
    fun `button click notification parses correctly`() {
        val payload = mapOf(
            "type" to "button_click",
            "title" to "Button Clicked",
            "body" to "John clicked 'Morning Routine'",
            "data" to mapOf(
                "button_id" to "btn-123",
                "button_name" to "Morning Routine",
                "user_id" to "user-456",
                "user_name" to "John"
            )
        )

        val notification = parseNotificationPayload(payload)

        assertThat(notification.type).isEqualTo("button_click")
        assertThat(notification.title).isEqualTo("Button Clicked")
        assertThat(notification.body).isEqualTo("John clicked 'Morning Routine'")
        assertThat(notification.data?.get("button_id")).isEqualTo("btn-123")
    }

    @Test
    fun `friend request notification parses correctly`() {
        val payload = mapOf(
            "type" to "friend_request",
            "title" to "Friend Request",
            "body" to "Jane sent you a friend request",
            "data" to mapOf(
                "friend_id" to "friend-789",
                "user_id" to "user-456",
                "user_name" to "Jane"
            )
        )

        val notification = parseNotificationPayload(payload)

        assertThat(notification.type).isEqualTo("friend_request")
        assertThat(notification.title).isEqualTo("Friend Request")
        assertThat(notification.data?.get("friend_id")).isEqualTo("friend-789")
    }

    @Test
    fun `system notification parses correctly`() {
        val payload = mapOf(
            "type" to "system",
            "title" to "ButtonLog",
            "body" to "Your subscription has been renewed",
            "data" to mapOf(
                "message_type" to "subscription_renewed"
            )
        )

        val notification = parseNotificationPayload(payload)

        assertThat(notification.type).isEqualTo("system")
        assertThat(notification.data?.get("message_type")).isEqualTo("subscription_renewed")
    }

    @Test
    fun `notification with missing fields has defaults`() {
        val payload = mapOf(
            "type" to "button_click"
            // Missing title, body, and data
        )

        val notification = parseNotificationPayload(payload)

        assertThat(notification.type).isEqualTo("button_click")
        assertThat(notification.title).isEmpty()
        assertThat(notification.body).isEmpty()
        assertThat(notification.data).isNull()
    }

    // MARK: - Deep Link Extraction

    @Test
    fun `button click deep link extracts correctly`() {
        val notificationData = mapOf("button_id" to "btn-123")

        val deepLink = extractDeepLink("button_click", notificationData)

        assertThat(deepLink).isEqualTo("buttonlog://button/btn-123")
    }

    @Test
    fun `friend request deep link extracts correctly`() {
        val notificationData = mapOf("friend_id" to "friend-456")

        val deepLink = extractDeepLink("friend_request", notificationData)

        assertThat(deepLink).isEqualTo("buttonlog://friends/requests")
    }

    @Test
    fun `friend accepted deep link goes to friends`() {
        val notificationData = mapOf("friend_id" to "friend-456")

        val deepLink = extractDeepLink("friend_accepted", notificationData)

        assertThat(deepLink).isEqualTo("buttonlog://friends")
    }

    @Test
    fun `unknown type deep link returns home`() {
        val deepLink = extractDeepLink("unknown", emptyMap())

        assertThat(deepLink).isEqualTo("buttonlog://home")
    }

    // MARK: - Notification Channel Tests

    @Test
    fun `notification channel IDs are correct`() {
        assertThat(NotificationChannels.BUTTON_ALERTS.id).isEqualTo("button_alerts")
        assertThat(NotificationChannels.FRIEND_REQUESTS.id).isEqualTo("friend_requests")
        assertThat(NotificationChannels.SYSTEM.id).isEqualTo("system")
    }

    @Test
    fun `notification type maps to correct channel`() {
        assertThat(getChannelForType("button_click")).isEqualTo(NotificationChannels.BUTTON_ALERTS)
        assertThat(getChannelForType("friend_request")).isEqualTo(NotificationChannels.FRIEND_REQUESTS)
        assertThat(getChannelForType("friend_accepted")).isEqualTo(NotificationChannels.FRIEND_REQUESTS)
        assertThat(getChannelForType("system")).isEqualTo(NotificationChannels.SYSTEM)
        assertThat(getChannelForType("unknown")).isEqualTo(NotificationChannels.SYSTEM)
    }

    // MARK: - Notification Grouping

    @Test
    fun `notifications group by type`() {
        val notifications = listOf(
            ParsedNotification("button_click", "Click 1", "", mapOf("button_id" to "btn-1")),
            ParsedNotification("button_click", "Click 2", "", mapOf("button_id" to "btn-1")),
            ParsedNotification("friend_request", "Request", "", emptyMap())
        )

        val grouped = groupNotifications(notifications)

        assertThat(grouped["button_click"]?.size).isEqualTo(2)
        assertThat(grouped["friend_request"]?.size).isEqualTo(1)
    }

    @Test
    fun `button click notifications group by button`() {
        val notifications = listOf(
            ParsedNotification("button_click", "Click 1", "", mapOf("button_id" to "btn-1")),
            ParsedNotification("button_click", "Click 2", "", mapOf("button_id" to "btn-1")),
            ParsedNotification("button_click", "Click 3", "", mapOf("button_id" to "btn-2"))
        )

        val grouped = groupButtonNotifications(notifications)

        assertThat(grouped["btn-1"]?.size).isEqualTo(2)
        assertThat(grouped["btn-2"]?.size).isEqualTo(1)
    }

    // MARK: - Notification Priority Tests

    @Test
    fun `button click has high priority`() {
        val priority = getPriority("button_click")
        assertThat(priority).isEqualTo(NotificationPriority.HIGH)
    }

    @Test
    fun `friend request has default priority`() {
        val priority = getPriority("friend_request")
        assertThat(priority).isEqualTo(NotificationPriority.DEFAULT)
    }

    @Test
    fun `system notification has low priority`() {
        val priority = getPriority("system")
        assertThat(priority).isEqualTo(NotificationPriority.LOW)
    }

    // MARK: - FCM Token Handling

    @Test
    fun `FCM token is validated correctly`() {
        val validToken = "fMxwX-example-token-123456789:APA91b-valid-token"
        val invalidToken = ""
        val shortToken = "abc"

        assertThat(isValidFCMToken(validToken)).isTrue()
        assertThat(isValidFCMToken(invalidToken)).isFalse()
        assertThat(isValidFCMToken(shortToken)).isFalse()
    }

    // MARK: - Notification Actions

    @Test
    fun `button click action is correct`() {
        val action = getActionForType("button_click")
        assertThat(action).isEqualTo("VIEW_BUTTON")
    }

    @Test
    fun `friend request has accept and decline actions`() {
        val actions = getActionsForType("friend_request")
        assertThat(actions).containsExactly("ACCEPT", "DECLINE")
    }

    // MARK: - Helper Methods

    private fun parseNotificationPayload(payload: Map<String, Any?>): ParsedNotification {
        val type = payload["type"] as? String ?: "unknown"
        val title = payload["title"] as? String ?: ""
        val body = payload["body"] as? String ?: ""
        @Suppress("UNCHECKED_CAST")
        val data = payload["data"] as? Map<String, String>

        return ParsedNotification(type, title, body, data)
    }

    private fun extractDeepLink(type: String, data: Map<String, String>): String {
        return when (type) {
            "button_click" -> {
                val buttonId = data["button_id"]
                if (buttonId != null) "buttonlog://button/$buttonId"
                else "buttonlog://buttons"
            }
            "friend_request" -> "buttonlog://friends/requests"
            "friend_accepted" -> "buttonlog://friends"
            else -> "buttonlog://home"
        }
    }

    private fun getChannelForType(type: String): NotificationChannels {
        return when (type) {
            "button_click" -> NotificationChannels.BUTTON_ALERTS
            "friend_request", "friend_accepted" -> NotificationChannels.FRIEND_REQUESTS
            else -> NotificationChannels.SYSTEM
        }
    }

    private fun groupNotifications(notifications: List<ParsedNotification>): Map<String, List<ParsedNotification>> {
        return notifications.groupBy { it.type }
    }

    private fun groupButtonNotifications(notifications: List<ParsedNotification>): Map<String, List<ParsedNotification>> {
        return notifications
            .filter { it.type == "button_click" }
            .groupBy { it.data?.get("button_id") ?: "unknown" }
    }

    private fun getPriority(type: String): NotificationPriority {
        return when (type) {
            "button_click" -> NotificationPriority.HIGH
            "friend_request", "friend_accepted" -> NotificationPriority.DEFAULT
            else -> NotificationPriority.LOW
        }
    }

    private fun isValidFCMToken(token: String): Boolean {
        return token.isNotBlank() && token.length > 10
    }

    private fun getActionForType(type: String): String {
        return when (type) {
            "button_click" -> "VIEW_BUTTON"
            "friend_request" -> "VIEW_REQUEST"
            else -> "VIEW"
        }
    }

    private fun getActionsForType(type: String): List<String> {
        return when (type) {
            "friend_request" -> listOf("ACCEPT", "DECLINE")
            else -> emptyList()
        }
    }
}

// MARK: - Data Classes

data class ParsedNotification(
    val type: String,
    val title: String,
    val body: String,
    val data: Map<String, String>?
)

enum class NotificationChannels(val id: String) {
    BUTTON_ALERTS("button_alerts"),
    FRIEND_REQUESTS("friend_requests"),
    SYSTEM("system")
}

enum class NotificationPriority {
    LOW, DEFAULT, HIGH
}
